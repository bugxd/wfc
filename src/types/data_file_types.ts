export interface DataTile {
  svg: string;
  north: string;
  east: string;
  south: string;
  west: string;
  rotiationDegrees: number[];
  frequency: number;
}

export interface DataFile {
  cellSize: number,
  tiles: DataTile[],
}
