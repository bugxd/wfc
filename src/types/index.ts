export interface Tile {
  svg: string;
  north: string;
  east: string;
  south: string;
  west: string;
}
export type Tiles = Tile[];

export interface Adjacent {
  tileId: number;
  north: number[];
  east: number[];
  south: number[];
  west: number[];
}

export type Adjacents = Adjacent[];

export interface EntropyCell {
  row: number;
  col: number;
  entropy: number;
}

export interface TileRemoval {
  row: number;
  col: number;
  tileId: number;
}

export enum Direction {
  NORTH = 0,
  EAST = 1,
  SOUTH = 2,
  WEST = 3,
}

export interface Neighbor {
  direction: Direction,
  row: number,
  col: number
}
export type Neighbors = Neighbor[];

export interface DataFile {
  cellSize: number,
  tiles: Tile[],
  frequencies: number[],
}

export interface TileEnablerCount  {
  tileId: number;
  north: number;
  east: number;
  south: number;
  west: number;
}

export interface TileCell {
  row: number,
  col: number,
  possible: number,
  svg: string | undefined,
}
