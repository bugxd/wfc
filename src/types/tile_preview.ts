export function getPreviewImage(cellSize: number, svg: string, north: string, east: string, south: string, west: string, degree: number): string {

  /**
   * g = group - all dynamic elements in a Group
   * transform translate - 20 20 offset to be in the middle of the text
   * transform rotate - rotate around the given center point round(cellSize/2) for given degree
   */
  return `<svg viewBox='0 0 ${cellSize+40} ${cellSize+40}' height='${cellSize+40}' width='${cellSize+40}' xmlns='http://www.w3.org/2000/svg'>
    <text x="${Math.round(cellSize/2)}" y="10" class="small">${north}</text>
    <text x="${cellSize+30}" y="${Math.round(cellSize/2)+10}" class="small" style="writing-mode: tb;">${east}</text>
    <text x="${Math.round(cellSize/2)}" y="${cellSize+30}" class="small">${south}</text>
    <text x="10" y="${Math.round(cellSize/2)+10}" class="small" style="writing-mode: tb;">${west}</text>
    <rect x="19" y="19" width="${cellSize+2}" height="${cellSize+2}" style="fill:rgb(255,255,255);stroke:rgb(255,0,0);stroke-width:1" />
    <g x='20' y='20' height='${cellSize}' width='${cellSize}' transform='translate(20, 20) rotate(${degree},${Math.round(cellSize/2)},${Math.round(cellSize/2)})'>
      ${svg}
    </g>
  </svg>`
}
