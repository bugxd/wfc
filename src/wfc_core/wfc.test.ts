import { Adjacents, Tiles } from "../types";
import Wfc from "./wfc";

describe("wfc tests", () => {
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.234567891);
  });

  afterEach(() => {
      jest.spyOn(global.Math, 'random').mockRestore();
  });

  it("should updateHeap", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];

    const sut = new Wfc({
      gridRows: 3,
      gridCols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents,
    });

    sut.updateHeap();

    expect(sut.heap).toMatchSnapshot();
  });

  it("should map grid", () => {
    const tiles: Tiles = [
      {
        svg: "+",
        north: "aba",
        east: "aba",
        south: "aba",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "|",
        north: "aba",
        east: "aaa",
        south: "aba",
        west: "aaa",
        rotiationDegree: 0,
        uid: "a",
      },
      {
        svg: "-",
        north: "aaa",
        east: "aba",
        south: "aaa",
        west: "aba",
        rotiationDegree: 0,
        uid: "a",
      }
    ];
    const frequencies: number[] = [1, 1, 1];
    const adjacents: Adjacents = [
      {
        tileId: 0,
        north: [0, 1],
        east: [0, 2],
        south: [0, 1],
        west: [0, 2],
      },
      {
        tileId: 1,
        north: [0, 1],
        east: [0],
        south: [0, 1],
        west: [0],
      },
      {
        tileId: 2,
        north: [0],
        east: [0, 2],
        south: [0],
        west: [0, 2],
      }
    ];

    const sut = new Wfc({
      gridRows: 3,
      gridCols: 3,
      tiles: tiles,
      frequencies: frequencies,
      adjacents: adjacents,
    });

    const result = sut.mapGrid();

    expect(result).toMatchSnapshot();
  });
})
