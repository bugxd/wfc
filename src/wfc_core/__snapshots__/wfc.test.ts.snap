// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wfc tests should map grid 1`] = `
Array [
  Array [
    Object {
      "col": 0,
      "possible": 3,
      "row": 0,
      "svg": undefined,
    },
    Object {
      "col": 1,
      "possible": 3,
      "row": 0,
      "svg": undefined,
    },
    Object {
      "col": 2,
      "possible": 3,
      "row": 0,
      "svg": undefined,
    },
  ],
  Array [
    Object {
      "col": 0,
      "possible": 3,
      "row": 1,
      "svg": undefined,
    },
    Object {
      "col": 1,
      "possible": 3,
      "row": 1,
      "svg": undefined,
    },
    Object {
      "col": 2,
      "possible": 3,
      "row": 1,
      "svg": undefined,
    },
  ],
  Array [
    Object {
      "col": 0,
      "possible": 3,
      "row": 2,
      "svg": undefined,
    },
    Object {
      "col": 1,
      "possible": 3,
      "row": 2,
      "svg": undefined,
    },
    Object {
      "col": 2,
      "possible": 3,
      "row": 2,
      "svg": undefined,
    },
  ],
]
`;

exports[`wfc tests should updateHeap 1`] = `
Heap {
  "heap": Array [
    Object {
      "col": 2,
      "entropy": 1.0986199429891996,
      "row": 2,
    },
    Object {
      "col": 1,
      "entropy": 1.0986199429891996,
      "row": 2,
    },
    Object {
      "col": 0,
      "entropy": 1.0986199429891996,
      "row": 2,
    },
    Object {
      "col": 2,
      "entropy": 1.0986199429891996,
      "row": 1,
    },
    Object {
      "col": 1,
      "entropy": 1.0986199429891996,
      "row": 1,
    },
    Object {
      "col": 0,
      "entropy": 1.0986199429891996,
      "row": 1,
    },
    Object {
      "col": 2,
      "entropy": 1.0986199429891996,
      "row": 0,
    },
    Object {
      "col": 1,
      "entropy": 1.0986199429891996,
      "row": 0,
    },
    Object {
      "col": 0,
      "entropy": 1.0986199429891996,
      "row": 0,
    },
  ],
}
`;
